vault/resource_aws_auth_backend_client.go

```diff
@@ -0,0 +1,157 @@
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"backend": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Unique name of the auth backend to configure.",
+				ForceNew:    true,
+				Default:     "aws",
+				// standardise on no beginning or trailing slashes
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+			"access_key": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "AWS Access key with permissions to query AWS APIs.",
+				Sensitive:   true,
+			},
+			"secret_key": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "AWS Secret key with permissions to query AWS APIs.",
+				Sensitive:   true,
+			},
+			"ec2_endpoint": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "URL to override the default generated endpoint for making AWS EC2 API calls.",
+			},
+			"iam_endpoint": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "URL to override the default generated endpoint for making AWS IAM API calls.",
+			},
+			"sts_endpoint": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "URL to override the default generated endpoint for making AWS STS API calls.",
+			},
+			"iam_server_id_header_value": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The value to require in the X-Vault-AWS-IAM-Server-ID header as part of GetCallerIdentity requests that are used in the iam auth method.",
+			},
+		},
+	}
+}
+
+func awsAuthBackendWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	// if backend comes from the config, it won't have the StateFunc
+	// applied yet, so we need to apply it again.
+	backend := strings.Trim(d.Get("backend").(string), "/")
+	accessKey := d.Get("access_key").(string)
+	secretKey := d.Get("secret_key").(string)
+	ec2Endpoint := d.Get("ec2_endpoint").(string)
+	iamEndpoint := d.Get("iam_endpoint").(string)
+	stsEndpoint := d.Get("sts_endpoint").(string)
+	iamServerIDHeaderValue := d.Get("iam_server_id_header_value").(string)
+
+	data := map[string]interface{}{
+		"access_key":                 accessKey,
+		"secret_key":                 secretKey,
+		"endpoint":                   ec2Endpoint,
+		"iam_endpoint":               iamEndpoint,
+		"sts_endpoint":               stsEndpoint,
+		"iam_server_id_header_value": iamServerIDHeaderValue,
+	}
+
+	log.Printf("[DEBUG] Writing AWS auth backend client config to %q", "/auth/"+backend+"/config/client")
+	_, err := client.Logical().Write("auth/"+backend+"/config/client", data)
+	if err != nil {
+		return fmt.Errorf("error writing to %q: %s", "auth/"+backend+"/config/client", err)
+	}
+	log.Printf("[DEBUG] Wrote AWS auth backend client config to %q", "auth/"+backend+"/config/client")
+
+	d.SetId("auth/" + backend + "/config/client")
```

Comment:
Probably want `"auth/" + backend + "/config/client"` to come from a convenience method at this point, it's hard-coded in 4 spots just right here 😄